### PR TITLE
Add serialization of samples.

### PR DESCRIPTION
**Pull request by tillahoffmann**
#### Submission Checklist

- [x] Run unit tests
- [x] Declare copyright holder and open-source license: see below

#### Summary

- add a call to `_assemble_draws` to the `__getstate__` magic function such that samples are included when a fit object is serialized with `pickle` or `dill`.
- remove checks before calling `_assemble_draws` because the function already performs this check internally.

#### Copyright and Licensing

Please list the copyright holder for the work you are submitting (this will be you or your assignee, such as a university or company): Harvard University

By submitting this pull request, the copyright holder is agreeing to license the submitted work under the following licenses:
- Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)

### CODE REVIEW

File: cmdstanpy/stanfit/mcmc.py [hunk 1]
```diff
@@ -129,6 +129,10 @@ def __getattr__(self, attr: str) -> np.ndarray:
             # pylint: disable=raise-missing-from
             raise AttributeError(*e.args)
 
+    def __getstate__(self) -> dict:
```
Comment (WardBrian) [inline cmdstanpy/stanfit/mcmc.py:132]: Would you mind adding a docstring or comment just explaining what this is used for/why it is here?

**Review by WardBrian — COMMENTED** [review]
This is very cool. Would you mind adding it to the non-MCMC objects?

**Comment by tillahoffmann** [comment]
So this requires a slightly hacky fix for the tests to work. I have renamed `test_mcmc_serialization` such that it is executed before `test_no_xarray`. The latter uses the `without_import` context manager which in turn uses `reload`. Using `reload` causes an inconsistent state because there are now two types with the same name (but they are two different types). See below for an example.

```python
def test_state(self):
    with self.without_import('xarray', cmdstanpy.stanfit.mcmc):
        pass
    # This will print `False`. Removing the context manager above ensures
    # the line prints `True`.
    print(CmdStanMCMC is cmdstanpy.stanfit.mcmc.CmdStanMCMC)
```

**Comment by tillahoffmann** [comment]
> This is very cool. Would you mind adding it to the non-MCMC objects?

I didn't check carefully, but they seemed to already work out of the box. Will double check though and add some tests.

**Comment by WardBrian** [comment]
We use pytest order (https://pypi.org/project/pytest-order/) to force test ordering for situations like that

**Pull request merged**